core/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java

```diff
@@ -88,4 +127,372 @@ public ScoreFunctionBuilder setMultiValueMode(String multiValueMode) {
         return this;
     }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (fieldName == null) {
+            validationException = addValidationError("field name cannot be null", validationException);
+        }
+        if (scale == null) {
+            validationException = addValidationError("scale must not be null", validationException);
+        }
+        if (decay <= 0 || decay >= 1.0) {
+            validationException = addValidationError("decay must be in range 0..1!", validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    protected ScoreFunction doScoreFunction(QueryShardContext parseContext) throws IOException {
+        // now, the field must exist, else we cannot read the value for the doc later
+        MappedFieldType fieldType = parseContext.fieldMapper(fieldName);
+        if (fieldType == null) {
+            throw new IllegalArgumentException("unknown field ["+fieldName+"]");
+        }
+        // dates and time need special handling
+        if (fieldType instanceof DateFieldMapper.DateFieldType) {
+            return handleDateVariable(parseContext, (DateFieldMapper.DateFieldType) fieldType);
+        } else if (fieldType instanceof GeoPointFieldMapper.GeoPointFieldType) {
+            return handleGeoVariable(parseContext, (GeoPointFieldMapper.GeoPointFieldType) fieldType);
+        } else if (fieldType instanceof NumberFieldMapper.NumberFieldType) {
+            return handleNumberVariable(parseContext, (NumberFieldMapper.NumberFieldType) fieldType);
+        } else {
+            throw new IllegalArgumentException("field ["+fieldName+"] is of type ["+fieldType+"], but only numeric types are supported.");
+        }
+    }
+
+    private AbstractDistanceScoreFunction handleNumberVariable(
+            QueryShardContext parseContext, NumberFieldMapper.NumberFieldType fieldType) throws IOException {
+        // to be consistent with the parser
+        if (this.scale == null || this.origin == null) {
+            throw new IllegalArgumentException("both ["+DecayFunctionBuilder.SCALE+"] and ["+DecayFunctionBuilder.ORIGIN+"] " +
+                    "must be set for numeric fields.");
+        }
+        // handle field specific defaults
+        double origin = getAsDouble(this.origin, 0, "[{" + DecayFunctionBuilder.ORIGIN + "}] must be a double for numeric fields.");
+        double scale = getAsDouble(this.scale, 0, "[{" + DecayFunctionBuilder.SCALE + "}] must be a double for numeric fields.");
+        double offset = getAsDouble(this.offset, DEFAULT_OFFSET_NUMBER, "[{" + DecayFunctionBuilder.OFFSET + "}] must be a double for numeric fields.");;
+        // create the function
+        //NO COMMIT indexService needs to available in our parseContext test!
+        IndexNumericFieldData numericFieldData = parseContext.getForField(fieldType);
```

Comment:
Didn't check any deeper than reading the two lines above but this smells like it could be similar to https://github.com/elastic/elasticsearch/pull/13381